lib/scholar/linear/logistic_regression.ex

```diff
@@ -85,127 +83,110 @@ defmodule Scholar.Linear.LogisticRegression do
-        {coeff, bias} = update_coefficients(x, x_t, y_t, {coeff, bias}, learning_rate)
-        {iter + 1, x, learning_rate, iterations, x_t, y_t, coeff, bias}
-      end
-
-    %__MODULE__{coefficients: final_coeff, bias: final_bias, mode: :binary}
+    fit_n(x, y, NimbleOptions.validate!(opts, @opts_schema))
   end
 
-  # Multinomial logistic regression
+  # Logistic Regression training loop
 
-  defnp fit_multinomial(x, y, opts) do
+  defnp fit_n(x, y, opts) do
     {_m, n} = x.shape
     iterations = opts[:iterations]
     learning_rate = opts[:learning_rate]
     num_classes = opts[:num_classes]
-    one_hot = Scholar.Preprocessing.one_hot_encode(y, num_classes: num_classes)
-    x_t = Nx.transpose(x)
+    y = Scholar.Preprocessing.one_hot_encode(y, num_classes: num_classes)
 
-    {_, _, _, _, _, _, _, final_coeff} =
-      while {iter = 0, x, learning_rate, n, iterations, one_hot, x_t,
-             coeff = Nx.broadcast(Nx.tensor(0, type: Nx.type(x)), {n, num_classes})},
+    {_, _, _, _, _, _, final_coeff, final_bias} =
+      while {iter = 0, x, learning_rate, n, iterations, y,
+             coeff =
+               Nx.broadcast(
+                 Nx.tensor(1.0, type: Nx.Type.to_floating(Nx.type(x))),
+                 {n, num_classes}
+               ),
+             bias =
+               Nx.broadcast(Nx.tensor(0, type: Nx.Type.to_floating(Nx.type(x))), {num_classes})},
             iter < iterations do
-        coeff = update_coefficients_multinomial(x, x_t, one_hot, coeff, learning_rate)
-        {iter + 1, x, learning_rate, n, iterations, one_hot, x_t, coeff}
+        {coeff_grad, bias_grad} = grad_loss(coeff, bias, x, y)
+        coeff = coeff - learning_rate * coeff_grad
+        bias = bias - learning_rate * bias_grad
+        {iter + 1, x, learning_rate, n, iterations, y, coeff, bias}
       end
 
     %__MODULE__{
-      coefficients: final_coeff,
-      bias: Nx.tensor(0, type: Nx.type(x)),
-      mode: :multinomial
+      coefficients: Nx.transpose(final_coeff),
+      bias: final_bias
     }
   end
 
-  # Normalized softmax
-
-  defnp softmax(t) do
-    normalized_exp = (t - Nx.reduce_max(t, axes: [0], keep_axes: true)) |> Nx.exp()
-    normalized_exp / Nx.sum(normalized_exp, axes: [0])
+  defnp cross_entropy(log_probs, targets) do
+    target_class = Nx.argmax(targets, axis: 1)
+    nll = Nx.take_along_axis(log_probs, Nx.new_axis(target_class, 1), axis: 1)
+    -Nx.mean(nll)
   end
 
-  # Gradient descent for binary regression
-
-  defnp update_coefficients(x, x_t, y_t, {coeff, bias}, learning_rate) do
-    {m, _n} = x.shape
-
-    logit = 1 / (Nx.exp(-(Nx.dot(x, coeff) + bias)) + 1)
-
-    diff = Nx.reshape(logit - y_t, {m})
-
-    coeff_diff = Nx.dot(x_t, diff) / m
-
-    bias_diff = Nx.sum(diff) / m
-
-    new_coeff = coeff - coeff_diff * learning_rate
-    new_bias = bias - bias_diff * learning_rate
-
-    {new_coeff, new_bias}
+  defnp loss_fn(coeff, bias, xs, ys) do
+    probs = log_softmax(Nx.dot(xs, coeff) + bias)
```

Comment:
One thing I found in Axon is that it is much more stable to optimize the loss function which uses logits. You can avoid doing this recomputation by "caching" the logits as metadata when computing the softmax. This is what we do in Axon, and is also what they do in Keras. I think here since you have more control internally you can actually just avoid computing softmax altogether, compute the loss from logits (there is a cross_entropy_from_logits loss in Axon), and then only compute softmax in `predict_probabilities`

Reply:
@seanmor5 you mean something like this? 
https://github.com/msluszniak/scholar/blob/add_cross_entropy_from_logits/lib/scholar/linear/logistic_regression.ex#L120